infrastructure/async/src/test/java/tech/pegasys/teku/infrastructure/async/SafeFutureTest.java

```diff
@@ -1077,6 +1077,44 @@ public void getImmediately_notDone() {
         .hasMessageContaining("Expected result to be available immediately, but was not");
   }
 
+  @Test
+  public void firstSuccess_completesExceptionallyIfPassedListIsEmpty() {
+    final SafeFuture<String> result = SafeFuture.firstSuccess(List.of());
+    SafeFutureAssert.assertThatSafeFuture(result)
+        .isCompletedExceptionallyWith(IllegalArgumentException.class)
+        .hasMessage("The provided List of futures should not be empty");
+  }
+
+  @Test
+  public void firstSuccess_completesWithFirstSuccessfulResponse() {
+    final IllegalStateException exception = new IllegalStateException("oopsy");
+
+    final SafeFuture<String> result =
+        SafeFuture.firstSuccess(
+            List.of(
+                SafeFuture.completedFuture("foo"),
+                SafeFuture.completedFuture("bar"),
+                SafeFuture.failedFuture(exception)));
+
+    assertThat(result).isCompletedWithValue("foo");
+  }
+
+  @Test
+  public void firstSuccess_completesExceptionallyWhenAllFuturesFail() {
+    final IllegalStateException exception = new IllegalStateException("oopsy");
+
+    final SafeFuture<String> result =
+        SafeFuture.firstSuccess(
+            List.of(
+                SafeFuture.failedFuture(exception),
+                SafeFuture.failedFuture(exception),
+                SafeFuture.failedFuture(exception)));
+
+    SafeFutureAssert.assertThatSafeFuture(result)
+        .isCompletedExceptionallyWith(IllegalStateException.class)
+        .hasMessage("oopsy");
+  }
+
```

Comment:
I would add a test passing a non completed future in the list.
It will check that we are not waiting all of them to complete.

Reply:
good point, done